src/states/joining_peer.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{
+    adult::{Adult, AdultDetails},
+    bootstrapping_peer::BootstrappingPeer,
+    common::{Base, Bootstrapped, BootstrappedNotEstablished},
+};
+use crate::{
+    chain::GenesisPfxInfo,
+    config_handler,
+    error::{InterfaceError, RoutingError},
+    id::{FullId, PublicId},
+    messages::{DirectMessage, HopMessage, RoutingMessage},
+    outbox::EventBox,
+    peer_manager::{PeerManager, PeerState},
+    peer_map::PeerMap,
+    quic_p2p::{NodeInfo, Peer},
+    routing_message_filter::RoutingMessageFilter,
+    routing_table::Authority,
+    state_machine::{State, Transition},
+    time::Instant,
+    timer::Timer,
+    xor_name::XorName,
+    NetworkService,
+};
+use std::{
+    fmt::{self, Display, Formatter},
+    net::SocketAddr,
+    time::Duration,
+};
+
+// Time (in seconds) after which bootstrap is cancelled (and possibly retried).
+const JOIN_TIMEOUT: Duration = Duration::from_secs(120);
+
+// State of a node after bootstrapping, while joining a section
+pub struct JoiningPeer {
+    network_service: NetworkService,
+    routing_msg_filter: RoutingMessageFilter,
+    msg_backlog: Vec<RoutingMessage>,
+    full_id: FullId,
+    min_section_size: usize,
+    peer_map: PeerMap,
+    timer: Timer,
+    join_token: u64,
+}
+
+impl JoiningPeer {
+    pub fn new(
+        network_service: NetworkService,
+        full_id: FullId,
+        min_section_size: usize,
+        timer: Timer,
+        peer_map: PeerMap,
+        node_infos: Vec<NodeInfo>,
+    ) -> Self {
+        let join_token = timer.schedule(JOIN_TIMEOUT);
+
+        let mut joining_peer = Self {
+            network_service,
+            routing_msg_filter: RoutingMessageFilter::new(),
+            msg_backlog: vec![],
+            full_id,
+            min_section_size,
+            timer: timer,
+            peer_map,
+            join_token,
+        };
+
+        for node_info in node_infos {
+            joining_peer.send_join_request(node_info);
+        }
+        joining_peer
+    }
+
+    pub fn into_adult(
```

Comment:
Could also be into_elder (or Infant, but see earlier comment on Infant)

Reply:
I think we shouldn't transform joining peers directly into Elders. In order to become an Elder, a node needs to be voted into Parsec, and the vote only happens after the node becomes accepted into the section. So I would leave this part of the process as is: Joining → Infant/Adult (if only briefly) → Elder.